app/validators/all_details_on_order_validator.rb

```diff
@@ -0,0 +1,39 @@
+class AllDetailsOnOrderValidator
+
+  cattr_accessor :order_detail_validator_class
+
+  # Build a validator that will take an Order in its constructor and validate each
+  # child order detail using the order_detail_validator_class.
+  #
+  # Example:
+  # validator = AllDetailsOnOrder.build(NotePresenceValidator)
+  # validator.new(order).valid?
+  #
+  # This will run NotePresenceValidator.new(order_detail).valid? for each
+  # element of @order.order_details
+  def self.build(order_detail_validator_class)
+    Class.new(AllDetailsOnOrderValidator) do
+      self.order_detail_validator_class = order_detail_validator_class
+    end
+  end
+
+  # This class should only be initialized through the `.build` method
+  def initialize(order)
+    raise "Should not be initialized directly. Use the `.build` factory method instead" unless order_detail_validator_class
+    @order = order
+  end
```

Comment:
The way these validators get used is there's an array of classes, and we loop over each class and call `clazz.new(order).valid?`. `AllDetailsOnOrderValidator` is essentially an abstract class. The `build` method is used to create a new subclass of `AllDetailsOnOrderValidator` that has a reference to order detail validator class so it can use it internally later. This class doesn't get instantiated for a while.

I originally wanted to add a `def order_detail_validator_class` inside the new subclass, but I couldn't get it to reference the passed in `order_detail_validator_class`.

The other option I considered was varying levels of generic vs specific. One step along the way was something like this. That seemed something where generalization into a superclass felt right (though maybe premature).

```ruby
class AllDetailsNotePresenceValidator
  ...
  def valid?
    # Loop over everything so that all order details get `errors` applied to them
    invalid_orders = @order.order_details.reject do |od|
      NotePresenceValidator.new(od).valid?
    end

    invalid_orders.none?
  end
```

Does that make sense?